packages/zoe/src/cleanOfferRules.js

```diff
@@ -0,0 +1,66 @@
+import harden from '@agoric/harden';
+import { assert, details } from '@agoric/assert';
+
+// O(n^2)
+export const cleanOfferRules = (roleNames, amountMaths, offerRules) => {
+  const wantObj = offerRules.want || {};
+  const offerObj = offerRules.offer || {};
+  const exitObj = offerRules.exit;
+
+  const exitProperties = Object.getOwnPropertyNames(exitObj);
+  assert(
+    exitProperties.length === 1,
+    details`exit ${offerRules.exit} should only have one key`,
+  );
+  const exitKind = exitProperties[0];
+  const acceptableKinds = ['OnDemand', 'AfterDeadline', 'Waived'];
+  assert(
+    acceptableKinds.includes(exitKind),
+    details`exit rule ${exitKind} was not an acceptable kind`,
+  );
+
+  const newExitRule = { kind: exitKind, ...exitObj[exitKind] };
+
+  const wants = Object.getOwnPropertyNames(wantObj).map(roleName => ({
+    role: roleName,
```

Comment:
Nevermind that question. Found it. But it's not clear to me that the role names in the offer are validated against the role names from the terms.

Reply:
Right now the role names in the offer are validated against the terms only on the contract side. It seems like we could move that into Zoe, though.

Reply:
Hmm. The contract still wants to be able to enforce the kind of offer a particular seat would have, which includes the role names and also their place in the `want` and `offer` offerRules structure. So we cannot move all of the validation into Zoe because Zoe doesn't know what the contract wants in this regard. But, maybe we want Zoe to try to validate the role names at least, even if the effort might be duplicated by the contract

Reply:
The part that zoe needs to valid is that the alleged role name is a string that passes the regexp defining valid role name spellings. I don't want to zoe to accept and handle `__proto__` as if it is a role name. That should be rejected as soon as possible, so it is handled by as little code as possible.

Given that, I agree that it's better for the contract to do further validation, rather than zoe.

Reply:
If Zoe validates that each alleged role name from the offer rules matches one of the roleNames from the validated `roles` does that count? In other words, the ascii validation happens at the creation of the contractInstance when roles are defined, and then the offer rules roles must be one of those.

Reply:
Yes, that would be fine since it would be impossible for an invalid spelling of a role name to match. It would be strictly stricter.